src/ahocode/ahocode.py

```diff
@@ -0,0 +1,371 @@
+        - If the Automaton was created with
+          Automaton(ahocorasik.STORE_INTS) then the value is
+          optional. If provided it must be an integer, otherwise it
+          defaults to len(automaton) which is therefore the order
+          index in which keys are added to the trie.
+        - If the Automaton was created with
+          Automaton(ahocorasik.STORE_LENGTH) then associating a
+          value is not allowed - len(word) is saved automatically as
+          a value instead.
+
+        Calling add_word() invalidates all iterators only if the new
+        key did not exist in the trie so far (i.e. the method
+        returned True).
+        """
+
+        if self._finalized:
+            raise ValueError('KeywordTree has been finalized.' +
+                             ' No more keyword additions allowed')
+        original_keyword = key
+        keyword_size = len(key)
+        if self.value_type == STORE_LENGTH:
+            value = keyword_size
+        if len(key) <= 0:
+            return
+        current_state = self._zero_state
+        for char_index in range(keyword_size):
+            char = key[char_index]
+            try:
+                current_state = current_state.transitions[char]
+            except KeyError:
+                next_state = State(self._counter, parent=current_state,
+                                   symbol=char)
+                self._counter += 1
+                current_state.transitions[char] = next_state
+                current_state = next_state
+        current_state.value = value
+        if current_state.success:
+            return False
+        current_state.success = True
+        current_state.matched_keyword = original_keyword
+        return True
+
+    def clear(self):  # real signature unknown; restored from __doc__
+        """
+        clear()
+
+        Remove all keys from the trie. This method invalidates all
+        iterators.
+        """
+        self._zero_state = State(0)
+        self._counter = 1
+        self._finalized = False
+
+    def exists(self, key):  # real signature unknown; restored from __doc__
+        """
+        exists(key) -> boolean
+
+        Return True if the key is present in the trie. Same as using
+        the 'in' keyword.
+        """
+        if not self._finalized:
+            current_state = self._zero_state
+            for char in key:
+                try:
+                    current_state = current_state[char]
+                except KeyError:
+                    return False
+            return True
+        else:
+            self.search_all(key)
+
+    def iter(self, string, start=None, end=None, *args, **kwargs):
+        if not self._finalized:
+            raise ValueError('Automaton has not been finalized. No search allowed. Call make_automaton() first.')
+        zero_state = self._zero_state
+        current_state = zero_state
+        offset = 0
+        if (start and end) is not None:

```

Comment:
I think it is better to use `==` and `!=` operators for value comparisons instead of `is`. From this answer https://stackoverflow.com/questions/132988/is-there-a-difference-between-and-is:

> `==` is for value equality. Use it when you would like to know if two objects have the same value.
> `is` is for reference equality. Use it when you would like to know if two references refer to the same object.

Please review and revise all use of `is` in your code, unless you really are checking to see if two variables are the same object in memory.